tiledb/common/evaluator/test/unit_evaluator.cc

```diff
@@ -0,0 +1,57 @@
+/**
+ * @file   tiledb/common/evaluator/test/unit_evaluator.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "test/support/tdb_catch.h"
+#include "tiledb/common/dynamic_memory/dynamic_memory.h"
+#include "tiledb/common/evaluator/evaluator.h"
+
+using namespace tiledb::common;
+
+TEST_CASE("constructor", "[evaluator]") {
+  auto f = [](std::string key) {
+    return std::make_tuple(key.size(), std::make_shared<std::string>(key));
+  };
```

Comment:
This definition of `f` illustrates two aspects of the design I don't think are appropriate.
1. The allocation is the responsibility of the function, not of the evaluator. If that's the case, then the evaluators shouldn't use `shared_ptr` anywhere. They can be instantiated with `shared_ptr<T>` by the user, rather than with `<T>`.
2. The function is responsible for providing the size of the object. There are three evaluators, and size is used by only one of them. That's illustrating an important way that the design is flawed: the required definition doesn't match the general case.